File: scripts/set_version_number.js
#! /usr/bin/env node
/* eslint-disable no-console */

'use strict';

const path = require('path');
const replace = require('replace');

function updatePreCommitHookScript(version) {
  replace({
    regex: /--package-version=[\d.]+/,
    replacement: `--package-version=${version}`,
    paths: [path.join('scripts', 'git', 'pre-commit')],
    silent: true
  });
}

function updateInstallerScript(version) {
  replace({
    regex: /#define MyAppVersion "[\d.]+"/,
    replacement: `#define MyAppVersion "${version}"`,
    paths: [path.join('scripts', 'installer.iss')],
    silent: true
  });
}

function updateCppFile(version) {
  const file = path.join('src', 'gui', 'version.cpp.in');
  const versionParts = version.split('.');

  replace({
    regex: /Version::major = \d+;/,
    replacement: `Version::major = ${versionParts[0]};`,
    paths: [file],
    silent: true
  });

  replace({
    regex: /Version::minor = \d+;/,
    replacement: `Version::minor = ${versionParts[1]};`,
    paths: [file],
    silent: true
  });

  replace({
    regex: /Version::patch = \d+;/,
    replacement: `Version::patch = ${versionParts[2]};`,
    paths: [file],
    silent: true
  });
}

function updateResourceFiles(version) {
  const files = [path.join('src', 'gui', 'resource.rc')];
  const commaSeparatedVersion = version.replace(/\./g, ', ');

  replace({
    regex: /VERSION \d+, \d+, \d+/g,
    replacement: `VERSION ${commaSeparatedVersion}`,
    paths: files,
    silent: true
  });

  replace({
    regex: /Version", "\d+\.\d+\.\d+"/g,
    replacement: `Version", "${version}"`,
    paths: files,
    silent: true
  });
}

function main() {
  if (process.argv.length !== 3) {
    console.error(
      'Invalid number of arguments given. Only one argument (the new version number) is expected.'
    );
    process.exit(1);
  }

  const newVersion = process.argv[2];

  if (newVersion.split('.').length !== 3) {
    console.error('The version number must be a three-part semantic version.');
    process.exit(1);
  }

  updatePreCommitHookScript(newVersion);
  updateInstallerScript(newVersion);
  updateCppFile(newVersion);
  updateResourceFiles(newVersion);
}

main();

File: scripts/set_version_number.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
import os
import re

def replace_in_file(path, regex, replacement):
    regex = re.compile(regex)

    lines = []
    with open(path, encoding='utf8') as infile:
        for line in infile:
            lines.append(re.sub(regex, replacement, line))

    with open(path, 'w', encoding='utf8') as outfile:
        for line in lines:
            outfile.write(line)

def update_pre_commit_hook_script(path, version):
    replace_in_file(path, '--package-version=[\\d.]+', '--package-version={}'.format(version))

def update_installer_script(path, version):
    replace_in_file(path, '#define MyAppVersion "[\\d.]+"', '#define MyAppVersion "{}"'.format(version))

def update_cpp_file(path, version):
    version_parts = version.split('.')

    replace_in_file(path, 'LOOT_VERSION_MAJOR = \\d+;', 'LOOT_VERSION_MAJOR = {};'.format(version_parts[0]))
    replace_in_file(path, 'LOOT_VERSION_MINOR = \\d+;', 'LOOT_VERSION_MINOR = {};'.format(version_parts[1]))
    replace_in_file(path, 'LOOT_VERSION_PATCH = \\d+;', 'LOOT_VERSION_PATCH = {};'.format(version_parts[2]))

def update_resource_file(path, version):
    comma_separated_version = version.replace('.', ', ')

    replace_in_file(path, 'VERSION \\d+, \\d+, \\d+', 'VERSION {}'.format(comma_separated_version))
    replace_in_file(path, 'Version", "\\d+\\.\\d+\\.\\d+"', 'Version", "{}"'.format(version))

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description = 'Set the LOOT version number')
    parser.add_argument('version', nargs='+')

    arguments = parser.parse_args()

    if len(arguments.version) != 1:
        raise RuntimeError('Invalid number of arguments given. Only one argument (the new version number) is expected.')

    if len(arguments.version[0].split('.')) != 3:
        raise RuntimeError('The version number must be a three-part semantic version.')

    version = arguments.version[0]

    update_pre_commit_hook_script(
        os.path.join('scripts', 'git', 'hooks', 'pre-commit'),
        version
    )
    update_installer_script(
        os.path.join('scripts', 'installer.iss'),
        version
    )
    update_cpp_file(
        os.path.join('src', 'gui', 'version.h'),
        version
    )
    update_resource_file(
        os.path.join('src', 'gui', 'resource.rc'),
        version
    )
